### PR TITLE
[OpenThread] Do not re-add existing services

### PR DESCRIPTION
**Pull request by Damian-Nordic**
#### Problem
Currently, when a single DNS-SD service is updated, all services are removed and re-added. This approach is
acceptable in the case of mDNS-based implementations, but in the case of SRP it generates unnecessary traffic to the
SRP server. Also, it may generate SRP updates bigger than MTU which has only been addressed in OpenThread recently and
may result in SRP update parsing errors when using older OpenThread versions and the SRP server does not support IPv6 reassembly.

#### Change overview
Detect if a service is re-added with the same port, subtypes and TXT entries and do nothing in such a case.
Reduce the size of SRP debug logs to compensate flash increase in debug builds. 

#### Testing
Tested on nRF Connect Lock that adding 5 fabrics and opening a commissionable window succeeds.
Verified using nRF sniffer for 15.4 that only modified services are sent in an SRP update.

### CODE REVIEW

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp [hunk 1]
```diff
@@ -2057,10 +2064,46 @@ void GenericThreadStackManagerImpl_OpenThread<ImplClass>::OnSrpClientStateChange
 }
 
 template <class ImplClass>
-bool GenericThreadStackManagerImpl_OpenThread<ImplClass>::SrpClient::Service::Matches(const char * aInstanceName,
-                                                                                      const char * aName) const
+bool GenericThreadStackManagerImpl_OpenThread<ImplClass>::SrpClient::Service::Matches(const char * instanceName,
+                                                                                      const char * name) const
+{
+    return IsUsed() && (strcmp(mService.mInstanceName, instanceName) == 0) && (strcmp(mService.mName, name) == 0);
```
Comment (jmeg-sfy) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:2070]: do we a max for these instanceName and name ? is it guaranteed to have a null terminator ?

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp [hunk 2]
```diff
@@ -2057,10 +2064,46 @@ void GenericThreadStackManagerImpl_OpenThread<ImplClass>::OnSrpClientStateChange
...
+
+template <class ImplClass>
+bool GenericThreadStackManagerImpl_OpenThread<ImplClass>::SrpClient::Service::Matches(
+    const char * instanceName, const char * name, uint16_t port, const Span<const char * const> & subTypes,
+    const Span<const Dnssd::TextEntry> & txtEntries) const
 {
-    return IsUsed() && (strcmp(mService.mInstanceName, aInstanceName) == 0) && (strcmp(mService.mName, aName) == 0);
+    size_t myNumSubTypes = 0;
+
+    for (const char * const * mySubType = mService.mSubTypeLabels; mySubType != nullptr && *mySubType != nullptr; ++mySubType)
```
Comment (jmeg-sfy) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:2080]:
```suggestion
    for (const char * const * mySubType = mService.mSubTypeLabels; (mySubType != nullptr) && (*mySubType != nullptr); ++mySubType)
```

**Review by jmeg-sfy — APPROVED** [review]

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp (hunk 1 above)
Comment (Damian-Nordic) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:2070]: We have a max, but all strings passed to the DNS-SD platform layer are constructed internally by the common DNS-SD layer, so they should be safe.

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp [hunk 3]
```diff
@@ -1959,10 +1959,12 @@ void GenericThreadStackManagerImpl_OpenThread<ImplClass>::OnSrpClientNotificatio
                                                                                   const otSrpClientService * aRemovedServices,
                                                                                   void * aContext)
 {
+    const char * errorStr = nullptr;
+
     switch (aError)
     {
     case OT_ERROR_NONE: {
-        ChipLogDetail(DeviceLayer, "OnSrpClientNotification: Last requested operation completed successfully");
+        ChipLogDetail(DeviceLayer, "SRP update succeeded");
```
Comment (jmeg-sfy) [inline src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:1967]: why not using the same errorStr = myMsg ?

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp (hunk 2 above)
Comment (Damian-Nordic) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:2080]: I personally find redundant parentheses less readable, but I recall there was a discussion on slack and some people were in favor of this convention, so I'll change that :).

File: src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp (hunk 3 above)
Comment (Damian-Nordic) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:1967]: Because it's `ChipLogDetail`, not `ChipLogError`.

Comment (jmeg-sfy) [reply src/platform/OpenThread/GenericThreadStackManagerImpl_OpenThread.cpp:1967]: ah ok got it !

**Review by andy31415 — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Review by saurabhst — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Review by LuDuda — APPROVED** [review]

**Review by tcarmelveilleux — APPROVED** [review]

**Pull request merged**